src/test/SubscribableRequestQueue.test.ts

```diff
@@ -0,0 +1,261 @@
+      const result = await promise2;
+      expect(result).to.equal("foo");
+    });
+
+    it("does not cancel an underlying request if it has another subscription", async () => {
+      const queue = new SubscribableRequestQueue(1);
+      const id1 = queue.addSubscriber();
+      const id2 = queue.addSubscriber();
+      // `block` keeps the queue full, to keep `test` from running
+      const block = queue.addRequestToQueue("block", id1, () => delay(TIMEOUT, undefined));
+
+      const promise1 = queue.addRequestToQueue("test", id1, () => delay(LONG_TIMEOUT, "foo"));
+      const promise2 = queue.addRequestToQueue("test", id2, () => delay(LONG_TIMEOUT, "bar"));
+      expect(queue.hasRequest("test")).to.be.true;
+      expect(queue.requestRunning("test")).to.be.false;
+      expect(queue.isSubscribed(id1, "test")).to.be.true;
+      expect(queue.isSubscribed(id2, "test")).to.be.true;
+
+      queue.cancelRequest("test", id1);
+      expect(queue.hasRequest("test")).to.be.true;
+      expect(queue.isSubscribed(id1, "test")).to.be.false;
+      expect(queue.isSubscribed(id2, "test")).to.be.true;
+      expect(await isRejected(promise1)).to.be.true;
+
+      await block;
+      expect(queue.requestRunning("test")).to.be.true;
+      const result = await promise2;
+      expect(result).to.equal("foo");
+    });
+
+    it("cancels an underlying request if it is waiting and has no other subscriptions", async () => {
+      const queue = new SubscribableRequestQueue(1);
+      const id = queue.addSubscriber();
+      // `block` keeps the queue full, to keep `test` from running
+      queue.addRequestToQueue("block", id, () => delay(TIMEOUT, undefined));
+
+      const promise = queue.addRequestToQueue("test", id, () => delay(LONG_TIMEOUT, "foo"));
+      expect(queue.hasRequest("test")).to.be.true;
+      expect(queue.requestRunning("test")).to.be.false;
+      expect(queue.isSubscribed(id, "test")).to.be.true;
+
+      queue.cancelRequest("test", id);
+      expect(await isRejected(promise)).to.be.true;
+      expect(queue.hasRequest("test")).to.be.false;
+      expect(queue.isSubscribed(id, "test")).to.be.false;
+    });
+  });
+
+  describe("removeSubscriber", () => {
+    it("removes a subscriber", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+      expect(queue.hasSubscriber(id)).to.be.true;
+      queue.removeSubscriber(id);
+      expect(queue.hasSubscriber(id)).to.be.false;
+    });
+
+    it("cancels all subscriptions from a removed subscriber", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id1 = queue.addSubscriber();
+      const id2 = queue.addSubscriber();
+
+      const promise1 = queue.addRequestToQueue("test", id1, () => delay(TIMEOUT, "foo"));
+      const promise2 = queue.addRequestToQueue("test", id1, () => delay(TIMEOUT, "bar"));
+      const promise3 = queue.addRequestToQueue("test", id2, () => delay(TIMEOUT, "baz"));
```

Comment:
Should `promise1` and `promise2` have different keys here? `promise1` would get rejected right away by the second `addRequestToQueue` call, which means you'd only be testing that one subscription is cancelled? (vs. all subscriptions getting cancelled)

Reply:
The perils of co-authoring your tests with Copilot 🙃. Great catch, thank you!